go/pkg/ddapi/handler.go

```diff
@@ -0,0 +1,221 @@
+	if terr != nil {
+		// Most likely bad input (bad request).
+		logErrorEmit400(w, fmt.Errorf("bad request: error while translating JSON doc: %v", terr))
+		return
+	}
+
+	writeRequest := &prompb.WriteRequest{
+		Timeseries: promTimeSeriesFragments,
+	}
+
+	log.Debugf("Prom write request: %s", writeRequest)
+
+	// Serialize struct into protobuf message (byte sequence).
+	pbmsgbytes, perr := proto.Marshal(writeRequest)
+	if perr != nil {
+		logErrorEmit500(w, fmt.Errorf("error while constructing Prometheus protobuf message: %v", perr))
+		return
+	}
+
+	// Snappy-compress the byte sequence.
+	spbmsgbytes := snappy.Encode(nil, pbmsgbytes)
+	// log.Debugf("snappy-compressed pb msg: %s", spbmsgbytes)
+
+	ddcp.postPromWriteRequestAndHandleErrors(w, spbmsgbytes)
+
+	// Make the DD agent's HTTP client happy.
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusAccepted)
+	w.Write([]byte("{\"status\": \"ok\"}"))
+}
+
+/*
+Try to send the HTTP POST request to a Prometheus remote_write endpoint, as
+provided by the Cortex distributor/ingester system.
+
+Maybe change this approach to using a ReverseProxy
+object as we do in
+https://github.com/opstrace/opstrace/blob/3f405cd4baa709c5d624d8966b8e2820b28ea37f/go/pkg/middleware/proxy.go#L75
+?
+
+The challenge with this approach might be response translation -- after all, we
+may need to have more flexibility in translating Cortex responses for the DD
+agent.
+*/
+func (ddcp *DDCortexProxy) postPromWriteRequestAndHandleErrors(w http.ResponseWriter, spbmsgbytes []byte) error {
+	req, err := http.NewRequest(
+		http.MethodPost,
+		ddcp.remoteWriteURL,
+		bytes.NewBuffer(spbmsgbytes),
+	)
+
+	// In which cases does this hit in?
+	if err != nil {
+		return err
+	}
+
+	// Cortex's remote_write endpoint expects a snappy-compressed protobuf
+	// message. Be explicit about what's sent.
+	req.Header.Add("X-Prometheus-Remote-Write-Version", "0.1.0")
+	req.Header.Add("Content-Encoding", "snappy")
+	req.Header.Set("Content-Type", "application/x-protobuf")
+
+	// Specify Cortex tenant to insert to.
+	req.Header.Set("X-Scope-OrgID", ddcp.tenantName)
+
+	resp, reqerr := ddcp.rwHTTPClient.Do(req)
+
+	if reqerr != nil {
+		// Which kinds of errors are handled here? Probably all those cases
+		// where the request could not be written to the tcp conn. TODO: emit
+		// 50x indicating gateway error?  I assume this would handle all
+		// transport-related errors while trying to interact with the remote
+		// system. For timeouts, we should therefore emit a 504 Gateway
+		// Timeout.
+		logErrorEmit500(w, fmt.Errorf("error while interacting with remote_write endpoint: %v", reqerr))
+	}
+	defer resp.Body.Close()
+
+	log.Infof("cortex HTTP response code: %v", resp.StatusCode)
+
+	bodybytes, readerr := ioutil.ReadAll(resp.Body)
+
+	if readerr != nil {
+		logErrorEmit500(w, fmt.Errorf("error while reading upstream response: %v", readerr))
+		return nil
```

Comment:
Why return nil and not the readerr?

Reply:
Oh, you saw a confusing work in progress state here that satisfies the compiler, but not common sense :dagger: 

Yes, `nil` is the wrong signal. The goal for this function is to return an error to signal to the caller of this function that the request processing is to be terminated, and that an error http response has already been written to the client!

Will clean that up, and also add comments.